src/gui/setuserstatusdialogmodel.h

```diff
@@ -0,0 +1,144 @@
+
+    explicit SetUserStatusDialogModel(std::shared_ptr<UserStatusConnector> userStatusJob,
+        QObject *parent = nullptr);
+
+    explicit SetUserStatusDialogModel(std::shared_ptr<UserStatusConnector> userStatusJob,
+        std::unique_ptr<DateTimeProvider> dateTimeProvider,
+        QObject *parent = nullptr);
+
+    explicit SetUserStatusDialogModel(const UserStatus &userStatus,
+        std::unique_ptr<DateTimeProvider> dateTimeProvider,
+        QObject *parent = nullptr);
+
+    explicit SetUserStatusDialogModel(const UserStatus &userStatus,
+        QObject *parent = nullptr);
+
+    ~SetUserStatusDialogModel();
+
+    Q_REQUIRED_RESULT UserStatus::OnlineStatus onlineStatus() const;
+    Q_INVOKABLE void setOnlineStatus(OCC::UserStatus::OnlineStatus status);
+
+    Q_REQUIRED_RESULT QUrl onlineIcon() const;
+    Q_REQUIRED_RESULT QUrl awayIcon() const;
+    Q_REQUIRED_RESULT QUrl dndIcon() const;
+    Q_REQUIRED_RESULT QUrl invisibleIcon() const;
+
+    Q_REQUIRED_RESULT QString userStatusMessage() const;
+    Q_INVOKABLE void setUserStatusMessage(const QString &message);
+    void setUserStatusEmoji(const QString &emoji);
+    Q_REQUIRED_RESULT QString userStatusEmoji() const;
+
+    Q_INVOKABLE void setUserStatus();
+    Q_INVOKABLE void clearUserStatus();
+
+    Q_REQUIRED_RESULT int predefinedStatusesCount() const;
+    Q_INVOKABLE UserStatus predefinedStatus(int index) const;
+    Q_INVOKABLE QString predefinedStatusClearAt(int index) const;
+    Q_INVOKABLE void setPredefinedStatus(int index);
+
+    Q_REQUIRED_RESULT QStringList clearAtStages() const;
+    Q_REQUIRED_RESULT QString clearAt() const;
+    Q_INVOKABLE void setClearAt(int index);
+
+    Q_REQUIRED_RESULT QString errorMessage() const;
+
+signals:
+    void errorMessageChanged();
+    void userStatusChanged();
+    void onlineStatusChanged();
+    void clearAtChanged();
+    void predefinedStatusesChanged();
+    void finished();
+
+private:
+    enum class ClearStageType {
+        DontClear,
+        HalfHour,
+        OneHour,
+        FourHour,
+        Today,
+        Week
+    };
+
+    void init();
+    void onUserStatusFetched(const UserStatus &userStatus);
+    void onPredefinedStatusesFetched(const std::vector<UserStatus> &statuses);
+    void onUserStatusSet();
+    void onMessageCleared();
+    void onError(UserStatusConnector::Error error);
+
+    Q_REQUIRED_RESULT QString clearAtStageToString(ClearStageType stage) const;
+    Q_REQUIRED_RESULT QString clearAtReadable(const Optional<ClearAt> &clearAt) const;
+    Q_REQUIRED_RESULT QString timeDifferenceToString(int differenceSecs) const;
+    Q_REQUIRED_RESULT Optional<ClearAt> clearStageTypeToDateTime(ClearStageType type) const;
+    void setError(const QString &reason);
+    void clearError();
+
+    std::shared_ptr<UserStatusConnector> _userStatusJob {};
+    std::vector<UserStatus> _predefinedStatuses;
+    UserStatus _userStatus;
+    std::unique_ptr<DateTimeProvider> _dateTimeProvider;
```

Comment:
probably should be an instance rather than a pointer

Reply:
How then use injection in the tests?